<a href="https://colab.research.google.com/github/ElofssonLab/kb8029-book/blob/main/notebooks/day07-discussion-1.ipynb" style="display:inline-block;padding:10px 18px;background-color:#F9AB00;color:#000000;font-weight:bold;text-decoration:none;border-radius:6px;font-family:sans-serif;font-size:14px;">&#9654;&nbsp; Open in Google Colab</a>

# Day 7 — In-class discussion problem (1 of 3)

Work this out **by hand in your group first** — then run the code cell
to check your answer before presenting.

## Find weights for a perceptron that implements OR

Using the {-1, +1} encoding (-1 = false, +1 = true) and a sign
activation (output +1 if z > 0, else -1), the OR truth table is:

| x1 | x2 | OR |
|---|---|---|
| -1 | -1 | -1 |
| -1 | +1 | +1 |
| +1 | -1 | +1 |
| +1 | +1 | +1 |

**As a group:**

1. Try to find weights $w_1, w_2$ and a bias $b$ so that
   $z = w_1 x_1 + w_2 x_2 + b$ reproduces this table through the sign
   activation. (Hint: think geometrically — where does the separating
   line need to sit relative to the four corners?)
2. Is your solution the *only* one that works, or are there others?

Run the cell below to check your weights (edit `w1, w2, b` to try your
own answer first).

In [1]:
def sign(z):
    return 1 if z > 0 else -1

w1, w2, b = 1.0, 1.0, 0.5   # <-- try your own group's answer here first

inputs = [(-1, -1), (-1, 1), (1, -1), (1, 1)]
expected = [-1, 1, 1, 1]

for (x1, x2), exp in zip(inputs, expected):
    z = w1 * x1 + w2 * x2 + b
    out = sign(z)
    print(f"x=({x1:>2},{x2:>2})  z={z:>5.1f}  output={out:>2}  expected={exp:>2}  {'OK' if out == exp else 'WRONG'}")

x=(-1,-1)  z= -1.5  output=-1  expected=-1  OK
x=(-1, 1)  z=  0.5  output= 1  expected= 1  OK
x=( 1,-1)  z=  0.5  output= 1  expected= 1  OK
x=( 1, 1)  z=  2.5  output= 1  expected= 1  OK


**Discussion point:** $w_1 = w_2 = 1$, $b = 0.5$ works: the only
"false" corner is $(-1,-1)$, giving $z = -1.5$; every other corner
gives $z \ge 0.5$. Geometrically, the separating line just needs to cut
off that one corner from the other three — there's a whole *range* of
weights that do this (e.g. $w_1=w_2=2, b=1$ also works, since scaling
every weight and the bias by the same positive constant never changes
which side of zero $z$ falls on). This is the same "any positive
rescaling gives the same decision boundary" fact from Day 1's practice
problems, now showing up as "infinitely many correct answers," not
just one.